### DISCUSSION TITLE
Is there a method to get Tangent vector on each cell of element?

### DISCUSSION THREAD

**Discussion by DavoodDamircheli** [firedrakeproject/firedrake#2374] [Firedrake support]
Hi,
 I know that I can use 
FacetNormal: to Return facet normal function for given mesh. 

But, is it possible to get the tangent vector on interior edge of a cell in firedrake?
Thank you in advance.

**Comment by colinjcotter** [comment]
In 3D there is not a unique tangent vector. In 2D you can apply the perp operator ie
n = FacetNormal(mesh)
t = as_vector([-n[1], n[0]])

On 5 Mar 2022, at 08:46, Davood Damircheli ***@***.***> wrote:

﻿

Hi,
I know that I can use
FacetNormal: to Return facet normal function for given mesh.

But, is it possible to get the tangent vector on interior edge of a cell in firedrake?
Thank you in advance.

—
Reply to this email directly, view it on GitHub<https://github.com/firedrakeproject/firedrake/discussions/2374>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABOSV4UABZFXDDMVMNHZOY3U6MNMJANCNFSM5P7NA2LQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by wence-** [comment]
There are `CellEdgeVectors(mesh)`, which might be what you want:

```python
from ufl.classes import CellEdgeVectors

cev = CellEdgeVectors(mesh)

assemble(inner(cev, cev)*dx)
```

works for me.
